### DISCUSSION TITLE
Migrating from other library – a way to "wrap" query result?

### DISCUSSION THREAD

**Discussion by mb8z** [prisma/prisma#23232] [Q&A]
### Question

I'm considering migration from another library where it was possible to "wrap" results in a class, and when querying, use the result attribute, as well as other class method/attributes out of the box.

I see that `prisma` has the way to do it with the `.$extends` functionality, but that's not "ideal" for migration. Is there a way to  achieve this and keep the typings?

I was trying to achieve this with something like:

```ts
import { Prisma, Invitation as PrismaInvitation } from '@prisma/client';
import * as Types from './types';
import Invitation from './model';

class Invitation {
  constructor(private _data: PrismaInvitation) {
    Object.assign(this, _data);
  }
  
  protected get = () => {
    return this._data;
  };

  helloWorld() {
    const { myAttribute } = this.get(); // Example of retrieving the attribute
    return 'Hello, world!';
  }
}

type ExtendedInvitation = PrismaInvitation & Invitation;

const generateInvitation = (data: PrismaInvitation) => {
  return new Invitation(data: PrismaInvitation) as ExtendedInvitation;
}

const withInvitation () => Prisma.defineExtension((prisma) => {
  return prisma
    .$extends({
      name: 'invitation-result-extended',
      result: {
        invitation: {
          status: {
            needs: { accepted_at: true, sent_at: true },
            compute(r) {
              // My code here
            },
          },
      },
    })
    .$extends({
      name: 'invitation-model-base',
      model: {
        invitation: {
          async findFirst(args?: Prisma.InvitationFindFirstArgs) {
            const context = Prisma.getExtensionContext(this);
            const data = await context.findFirst(args);
            if (!data) return null;
            // TODO: Somehow wrap the data in the invitation and keep typings
            return generateInvitation(data);
          },
          // ... Overwritting other methods here
        },
      },
    });

export default withInvitation;
```

I'd then use this `withInvitation` to extend the client that I'd import in other files.

This "forcing" of types works "kind

**Comment by nurul3101** [comment]
Hey @mb8z 👋 

Could you please elaborate on what you meant by this statement?
> I see that prisma has the way to do it with the .$extends functionality, but that's not "ideal" for migration

Which migration are you referring to? Do you migration from a different library?

`$extends` allows you to add custom methods to models and results, as well as to modify query behavior. But, I don't think the direct integration of class-based extensions with Prisma's type system and the exact mechanism for wrapping query results in a custom class is directly covered.

**Reply by mb8z** [reply]
> Do you migration from a different library?

@nurul3101 Yes, exactly! I've been using Sequelize for quite some time and was wondering if there is a way to follow their practices.

**Reply by ludralph** [reply]
Hi @mb8z 

Prisma doesn't natively support this just like Nurul mentioned.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
